File: 347-top-k-frequent-elements/top-k-frequent-elements.java
import java.util.ArrayList;
import java.util.HashMap;

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        ArrayList<Integer>[] freq = new ArrayList[nums.length + 1];
        HashMap<Integer, Integer> count = new HashMap<>();

        // Count frequencies
        for (int num : nums) {
            count.put(num, count.getOrDefault(num, 0) + 1);
        }

        // Initialize each bucket
        for (int i = 0; i < freq.length; i++) {
            freq[i] = new ArrayList<>();
        }

        // Place numbers into frequency buckets
        count.forEach((n, c) -> {
            freq[c].add(n);
        });

        int[] res = new int[k];
        int idx = 0;

        // Traverse buckets from high frequency to low
        for (int i = freq.length - 1; i >= 0 && idx < k; i--) {
            for (int n : freq[i]) {
                res[idx++] = n;
                if (idx == k) {
                    return res;
                }
            }
        }

        return res; // in case k == 0 or other edge case
    }
}

File: 347-top-k-frequent-elements/top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = Counter(nums)

        min_heap = []

        for num, freq in count.items():
            heapq.heappush(min_heap, (freq, num))
            
            if len(min_heap) > k:
                heapq.heappop(min_heap)
        
        return [x[1] for x in min_heap]

        # time: O(m logk), m = # of unique numbers
        # space: O(m + k)